### DISCUSSION TITLE
Multiselect with control buttons

### DISCUSSION THREAD

**Discussion by marcospassos** [ariakit/ariakit#1547] [Questions] [answered]
I'm working on a [multiple selectbox component ](https://codesandbox.io/s/crimson-dawn-w8huqq) that requires buttons for un/selecting all options and applying changes. 

I found no way, however, to render the select popover and list separately – the popover is always rendered as the listbox itself. Setting `composite=false` makes the popover no longer receive the `role=listbox`, but other attributes like `aria-mutiselectable` are still passed, which breaks on screenreaders like VoiceOver. So I temporarily stripped out those attributes (which is probably a bad idea since new attributes returned from `useSelectPopover` may break it again in the future). However, I still get issues with focus: if no item is selected, the popover focuses on the first focusable child, which is the "Select all" element instead of the listbox. My next attempt was setting the `initialFocusRef`, but it overrides the default behavior causing the currently selected item to no longer be focused when opening the popover.

What is the proper way to render the popover and list individually? 

*Side question related to the use case:* is there any better way to reset the value on close (the current implementation is called multiple times since the `currentOptions` is in the dep list)?

**Comment by diegohaz** [comment, marked as answer]
> Setting `composite=false` makes the popover no longer receive the `role=listbox`, but other attributes like `aria-mutiselectable` are still passed, which breaks on screenreaders like VoiceOver.

That looks like a bug. Could you please file an issue?

> However, I still get issues with focus: if no item is selected, the popover focuses on the first focusable child, which is the "Select all" element instead of the listbox. My next attempt was setting the `initialFocusRef`, but it overrides the default behavior causing the currently selected item to no longer be focused when opening the popover.

You can conditionally pass `initialFocusRef`:

```jsx
const initialFocusRef = useRef(null);

<SelectPopover initialFocusRef={!currentValue.length ? initialFocusRef : undefined}>
  ...
  {list.map((value, i) => <SelectItem ref={i === 0 ? initialFocusRef : undefined} />}
  ...
</SelectPopover>
```

> _Side question related to the use case:_ is there any better way to reset the value on close (the current implementation is called multiple times since the `currentOptions` is in the dep list)?

The effect callback running multiple times shouldn't be a problem in that case since you're calling `select.setValue` with the same value. But that's being called multiple times because you've included `select` in the deps array. I understand that might've been a misleading suggestion of `react-hooks/exhaustive-deps`, but you generally shouldn't include the state object in the deps array.

Either ignore the eslint rule:

```jsx
useEffect(() => {
    if (!select.mounted) {
      if (isDirty(currentValue, select.value)) {
        select.setValue(currentValue);
      }
    }
  },
  // eslint-disable-next-line react-hooks/exhaustive-deps
  [select.mounted, currentValue, select.value, select.setValue]
);
```

Or do something like this:

```jsx
const { mounted, value, setValue } = select;

useEffect(() => {
    if (!mounted) {
      if (isDirty(currentValue, value)) {
        setValue(currentValue);
      }
    }
  },
  [mounted, currentValue, value, setValue]
);
```

**Reply by marcospassos** [reply]
> You can conditionally pass `initialFocusRef`:
> 
> ```js
> const initialFocusRef = useRef(null);
> 
> <SelectPopover initialFocusRef={!currentValue.length ? initialFocusRef : undefined}>
>   ...
>   {list.map((value, i) => <SelectItem ref={i === 0 ? initialFocusRef : undefined} />}
>   ...
> </SelectPopover>
> ```

My first try was something[ along these lines](https://codesandbox.io/s/inspiring-pine-g4f34u?file=/src/select-multiple.tsx). The problem I've found is that it causes a scroll shift:

https://user-images.githubusercontent.com/943036/175944901-73da1b10-d8f1-49c2-bb59-4feff5a9ba12.mov

_(notice it only happens after the first item is selected/unselected, no matter the position in the list)._

My guess is that the component tries to focus on the last enabled item on subsequent re-renders, causing an undesired shift, but I didn't dig deep.

**Reply by diegohaz** [reply]
Yeah, changing `initialFocusRef` while the dialog is open will move focus even if the dialog is already focused. This is necessary because the content within the dialog may be asynchronously rendered. The solution is to use a state that only changes when the dialog is hidden, like your `currentValue`, which I used in my suggestion.

**Reply by marcospassos** [reply]
I can't assume the current value is only updated on close since the user can apply whenever they want. It's a hybrid select + multiselect element, so not always the current value reflects the state when closed. I'd have to add an extra state just to make the focus work.  I'm wondering whether there is a simpler way.


Please let me know if you come up with a different idea :)

**Reply by marcospassos** [reply]
It gets even trickier if the component is controlled: besides a state, it will require a `useEffect` to work correctly.

**Reply by marcospassos** [reply]
@diegohaz, given that the `SelectPopover` is a specialization of the `Popover`, how about treating the [passed `initialFocusRef`](https://github.com/ariakit/ariakit/blob/729971c1471e3ccd16ece63cae568357f3741704/packages/ariakit/src/select/select-popover.ts#L43) as a fallback instead of the primary ref?

```diff
export const useSelectPopover = createHook<SelectPopoverOptions>(
-  ({ state, ...props }) => {
+  ({ state, initialFocusRef, ...props }) => {
    const values = toArray(state.value);
    const value = values[values.length - 1] ?? "";
    const [item, setItem] = useState<Item | null>(null);

    // Sets the initial focus ref.
    useEffect(() => {
      setItem((prevItem) => {
        if (state.mounted && prevItem?.ref.current) return prevItem;
        const item = findEnabledItemByValue(state.items, value);
        return item || null;
      });
    }, [state.mounted, state.items, value]);

    props = useSelectList({ state, ...props });
-    props = usePopover({ state, initialFocusRef: item?.ref, ...props });
+    props = usePopover({ state, initialFocusRef: item?.ref ?? initialFocusRef, ...props });

    return props;
  }
);
```

This change makes achieving the described behavior as easy as passing the fallback ref:

```tsx
<SelectPopover state={state} initialFocusRef={select.baseRef}>
...
</SelectPopover>
```

Alternatively:

```tsx
<SelectPopover state={state} fallbackFocusRef={select.baseRef} initialFocusRef={/* current behavior */}>
</SelectPopover>
```

I can submit a PR if you're ok with the proposed solution.

I've tried other alternatives, but duplicating the whole `useSelectPopover` seems the only viable alternative so far.

**Reply by diegohaz** [reply]
Can you open a feature request for this? There may be better APIs. We can discuss it there.

**Reply by marcospassos** [reply]
[Done](https://github.com/ariakit/ariakit/issues/1553)
